### ISSUE TITLE
Dragging pieces

### ISSUE THREAD

**Issue by optim-ally** [Infinite-Chess/infinitechess.org#22]
Dragging the board is a new kind of user interaction in infinite chess vs 8x8 chess. It makes the familiar action of dragging pieces in online chess problematic. I tried several time to drag pieces in infinite chess before remembering that it didn't work like that, but perhaps it could!

We could introduce ways to explicitly choose "piece dragging" or "board dragging", like right clicking or holding down a particular key, but I expect most users won't bother with those.

I propose the following logic for handling "drag" interactions:

1. if the board is already moving (e.g. it has remaining momentum from being dragged), then drag the board
2. else if the click action was a right-click or CMD/CTRL was held, drag the board
3. else if the cursor is not over a piece of the player's colour, drag the board
4. else if the cursor is over multiple pieces (when zoomed far enough out), drag the board
5. else drag the piece

This might need to go through several iterations if it's going to work, but what are initial thoughts?

**Comment by Heinrich-XIAO** [comment]
I feel like a good solution to this would be to drag the piece if the mouse was directly on the piece and drag the board otherwise.

**Comment by optim-ally** [comment]
Essentially, yes, but we need to be careful with the details. For example, if the board is zoomed out far enough then the mouse may be on top of multiple pieces.

Another issue is whether we consider the mouse "on" a piece if it is in the same square or if it actually needs to be inside the rendered shape of the piece. If it's the latter then cosmetic changes to the board and pieces could affect playing experience by making pieces harder or easier to move. I don't think that should be the case.

**Comment by Naviary2** [comment]
> 1. if the board is already moving (e.g. it has remaining momentum from being dragged), then drag the board

When dragging the board with the mouse and letting go, it no longer has momentum, it stops immediately. So it can't be *thrown* in this manner.

> if the board is zoomed out far enough then the mouse may be on top of multiple pieces.

The coordinates of the mouse is a single point, so it can never technically be over multiple pieces, at least in the code. But at that zoom distance clicking and dragging probably shouldn't be selecting a piece anyway, the user probably intends on moving the board.

> I feel like a good solution to this would be to drag the piece if the mouse was directly on the piece and drag the board otherwise.

I like this solution and I have actually thought this before! When I check out Lichess.org's mouse controls, they also have click to select, and dragging as well! Should have thought of that when I was making the first release! I personally think this is the best solution. But it can be tested!

**Comment by optim-ally** [comment]
> When dragging the board with the mouse and letting go, it no longer has momentum, it stops immediately. So it can't be thrown in this manner.

I'm planning ahead to when we have #23 😛 

You pointed out on that issue that scrolling with W,A,S,D already gives the board "momentum". In that case, I think piece dragging should be disabled until the board has come to a halt.

**Comment by Naviary2** [comment]
Waiting for #23 Board momentum after dragging

**Comment by Entificationcurry** [comment]
NAVIARY ACCEPT MY PULL REQUEST TO RESOLVE #23

**Comment by Naviary2** [comment]
RIP 🪦

**Comment by BF5258** [comment]
My attempt at implementing it: https://github.com/BF5258/infinitechess.org/tree/dragging-pieces
I tried to make it work similar to Chess.com and Lichess.

Lichess deselects the piece if it is dropped on an invalid square. Chess.com keeps it selected. Which is preferred?

**Comment by Naviary2** [comment]
Sorry I didn't create this list sooner! Here's how I imagined the various behaviors of dragging pieces would go:

* When holding a piece, the piece snaps to the position of the mouse. It's normal rendering is blocked (exactly like how the animation.js script blocks the rendering of the normal piece), and existing methods are used to render it exactly where the mouse is.
* The act of dragging a piece always selects it. Even if you drop it on an invalid square, it remains selected. HOWEVER, if you drag a piece onto the same square it's on (pick up and let go in the same place), then it will *toggle* its selection (unselect if selected, select if unselected).
* On mobile, the piece being dragged needs to be shifted upward buy a constant amount of pixels so that your finger doesn't cover it up. It's size while being dragged also needs to be constant and not proportional to how big the squares are.

The behavior with arrow indicators is special:

* If you drag a piece onto a single arrow indicator pointing to an enemy piece off screen, it will make the capture if it's legal. If you drag it onto multiple arrow indicators overlapping, it chooses the one that is legal (I think there should only ever be 1 legal capture if 2 arrows are overlapping, right?) You can tell if it's about to make the capture because the arrow indicator will be replaced by the image of the piece you have selected.
* Jumping (or individual moves) are ignored when dragging onto arrow indicators, it only works if the piece you've selected can slide there.
* drag one of the arrow indicators pointing to one of your pieces off screen to automatically select it and slide it across the edge of the screen. For example, if you work is offscreen to the right, and you drag arrow indicator upward, it will move the rock upward.

**Comment by Naviary2** [comment]
> Lichess deselects the piece if it is dropped on an invalid square. Chess.com keeps it selected. Which is preferred?

Yes! I think I prefer chess.com's way, to keep it selected even if it's dropped on an invalid square. With such a big board, we don't know how far people are going to be dragging their pieces. If they accidentally drag it onto an invalid square, they shouldn't have to move their mouse all the way back to re-drag it a second time.

**Comment by BF5258** [comment]
> I think there should only ever be 1 legal capture if 2 arrows are overlapping, right?

What should it do when it's on the edge of the visible area?

![image](https://github.com/user-attachments/assets/1e16c24b-ab59-44a7-9be9-ccbe89b36088)

**Comment by Naviary2** [comment]
> What should it do when it's on the edge of the visible area?

In this scenario, I think we should prioritize clicking the arrow (and subsequently panning to it) instead of selecting the piece.

**Comment by Naviary2** [comment]
I would advice polishing the first half of dragging pieces, and getting that merged, before moving onto the more challenging interactions with the arrow indicators 🧐

**Comment by Naviary2** [comment]
(Removing the "waiting for other feature" tag)

**Comment by Naviary2** [comment]
@BF5258 Another thing is dragging pieces should be toggleable, so we can later add a settings option for disabling it. But that will be a separate PR.

Clicking to select, however, I think should always be on no matter what, as it'd be very hard to be forced to drag long distances.

**Comment by Naviary2** [comment]
The settings dropdown spoke of could be called "Selection", and will have 2 toggles for both "Drag Pieces" and "Premoving"

**Comment by Naviary2** [comment]
Phase one of dragging pieces is complete. Next up is the behavior of dragging with arrow indicators.
